Change minlength/maxlength behavior around linebreaks

domenic opened the pull request:
This is a successor to #1517 which concentrates on the less-controversial part of the change, of changing linebreaks to count as 1 character (aligning with Edge and Firefox and with @tkent-google's suggestion for Chrome).

/cc @tkent-google, @rniwa, @kojiishi. In particular I'm hoping to hear from @rniwa if he is willing to change this aspect of the behavior. Let's mark this as "do not merge yet" until he weighs in.

---

As discussed in #1467, the current situation around these attributes is
not very interoperable. Some browsers count line breaks as two
characters (per the spec before this change), others as one character.

Per discussions, this updates minlength and maxlength to count line
breaks as one character. We believe this is the most developer-friendly
approach, as evidenced in part by repeated complaints against Chromium
for its behavior following the previous standard.

While here, updated the textLength property to return the code-point
length of the element's API value, instead of the element's value, since
as per http://software.hixie.ch/utilities/js/live-dom-viewer/?saved=4304
that is what browsers actually do.

Fixes part of #1467, but the debate remains about code-unit length vs.
code-point length vs. number of grapheme clusters.

tkent-google commented:
lgtm though I'm not sure if we should convert "normalize" to "normalise" in this change.

rniwa commented:
Yeah, treating CRLF as a single character for the purpose of applying max/minlength seems like a good idea and it will improve the interoperability of browsers.

annevk commented on source:
```diff
@@ -49631,28 +49631,30 @@ interface <dfn>HTMLTextAreaElement</dfn> : <span>HTMLElement</span> {
-  transformation applied:</p>
+  <p>The algorithm for obtaining the element's <span data-x="concept-fe-api-value">API value</span>
+  is to return the element's <span data-x="concept-textarea-raw-value">raw value</span> with the
+  <span>textarea linebreak normalisation transformation</span> applied. The <dfn>textarea linebreak
```
24 hits for "line break" in HTML, 0 for "linebreak"

annevk commented on source:
```diff
@@ -49686,7 +49688,7 @@ interface <dfn>HTMLTextAreaElement</dfn> : <span>HTMLElement</span> {
   <p>If the <code>textarea</code> element has a <span>maximum allowed value length</span>, then the
   element's children must be such that the <span>code-unit length</span> of the value of the
-  element's <code>textContent</code> IDL attribute with the <span>textarea wrapping
+  element's <code>textContent</code> IDL attribute with the <span>textarea linebreak normalisation
```
The commit message says nothing about the wrapping transformation not being relevant here. Are you sure that should be removed?

domenic replied:
This is a conformance requirement that mismatched the actual check that maxlength does. Let me add a bit to the commit message.

annevk commented on source:
```diff
@@ -49631,28 +49631,30 @@ interface <dfn>HTMLTextAreaElement</dfn> : <span>HTMLElement</span> {
-  <p>The element's <dfn data-x="concept-textarea-api-value">API value</dfn> is defined to be the
-  element's <span data-x="concept-textarea-raw-value">raw value</span> with the following
-  transformation applied:</p>
+  <p>The algorithm for obtaining the element's <span data-x="concept-fe-api-value">API value</span>
```
Why would we apply this to all form elements and not just `textarea` elements?

annevk replied:
I see, I guess this is defined in the `textarea` element section and therefore only `textarea` overrides the default.

domenic commented:
Updated linebreak vs. line break. New suggested commit message including the bit about the conformance requirement:

```
Change minlength/maxlength behavior around linebreaks 

As discussed in #1467, the current situation around these attributes is
not very interoperable. Some browsers count line breaks as two
characters (per the spec before this change), others as one character.

Per discussions, this updates minlength and maxlength to count line
breaks as one character. We believe this is the most developer-friendly
approach, as evidenced in part by repeated complaints against Chromium
for its behavior following the previous standard.

While here, updated the textLength property to return the code-point
length of the element's API value, instead of the element's value, since
as per http://software.hixie.ch/utilities/js/live-dom-viewer/?saved=4304
that is what browsers actually do. Similarly, updated the conformance
requirement on the text content of the textarea element to match
maxlength's actually-implemented behavior.

Fixes part of #1467, but the debate remains about code-unit length vs.
code-point length vs. number of grapheme clusters.
```

annevk commented:
Thanks @domenic!
